### PR TITLE
[Performance] Add async exponential while model executing

### PR DESCRIPTION
**Pull request by YuhanBai**
### What this PR does / why we need it?
Add a control to enable the exponential distribution operator overlapping with model executing (default is OFF due to this feature might not perform well on MOE models, i.e. For Qwen3-30B).
Enable async exponential overlapping will provides performance improvement.
Also, overlapping the exponential operator with module execution can cover the performance drop introduced by AICPU-version's exponential operator.

**UPDATE**: (12/12)
Now our overlap will use the same stream that introduced in this pr: #4908 .
We move the `do_async_exponential` from `model_runner_v1.py` to `sampler.py`.
Now we are using `additional_config` to enable async exponential:
Add `"enable_async_exponential": 1` in `addition_config`.
Now we **ONLY** support default exponential/AI-CPU exponential, the old `"enable_async_exponential": 2` option has been aborted to keep consistency.

### Does this PR introduce _any_ user-facing change?
**YES**, added a new `additional_config` : `"enable_async_exponential": 1`.
When `enable_async_exponential` is set to 1, we enable the async exponential and overlap with model runner.
When `enable_async_exponential` is set to 0 (default is 0), we disable the async exponential, but exponential will still running on a different stream using stream introduced in #4908.

A server example:
```shell
export VLLM_USE_V1=1
export TASK_QUEUE_ENABLE=1
export HCCL_BUFFSIZE=700
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True

python3 -m vllm.entrypoints.openai.api_server \
    --model=/home/data/Qwen2.5-32B \
    --served-model-name auto \
    --port 8006 \
    -tp 4 \
    -dp 1 \
    --enforce-eager \
    --disable-log-requests \
    --max-num-seqs 1024 \
    --gpu-memory-utilization 0.9 \
    --additional_config='{"enable_shared_expert_dp": false, "enable_async_exponential": 1}'
```

### How was this patch tested?
A2 Qwen2.5-32B/A3 Qwen2.5-32B/A3 Qwen3-32B/A3 Qwen3-30B

In this PR, we test this feature on A2/A3 platform and on three different models: `Qwen2.5-32B`; `Qwen3-30B`; `Qwen3-32B`.

#### A2 Tests

On A2 platform, we only tested the `Qwen2.5-32B`:

| | test1 (tok/s)| test2 (tok/s) | test3 (tok/s) | test4 (tok/s) | test5 (tok/s) | AVG (tok/s) |
| --- | --- | --- | --- | --- | --- | --- |
| Disable overlap | 1934.59 | 2161.24 | 2155.04 | 2155.58 | 2162.43 |2,113.7|
| Enable overlap | 1972.98 | 2205.74 | 2210.58 | 2210.72 | 2209.36 |**2,161.8**|

> Tests were using these cofigs: tp4dp1 enforce-eager num_prompts=48, 2k->2k

On A2 platform, enable asnyc exponential overlap will provide about 2.28% performance improvment.

#### A3 Tests

On A3 platform, we tested all three models: `Qwen2.5-32B`; `Qwen3-30B`; `Qwen3-32B`
For `Qwen2.5-32B` and `Qwen3-32B`, enable the asnyc exponential overlap will give significant performance improvement:

**Qwen2.5-32B**

| | test1 (tok/s)| test2 (tok/s) | test3 (tok/s) | test4 (tok/s) | test5 (tok/s) | AVG (tok/s) |
| --- | --- | --- | --- | --- | --- | --- |
| Disable overlap | 1418.36 | 1632.18 | 1648.44 | 1670.25 | 1511.80|1576.21|
| Enable overlap | 1601.71 | 1675.43 | 1740.61 | 1720.02 | 1696.65|**1686.88**|

> Tests were using these cofigs: tp4dp1 enforce-eager num_prompts=48, 1.5k->2k

**Qwen3-32B**

| | test1 (tok/s)| test2 (tok/s) | test3 (tok/s) | test4 (tok/s) | test5 (tok/s) | AVG (tok/s) |
| --- | --- | --- | --- | --- | --- | --- |
| Disable overlap | 967.51| 1060.02| 1053.79|1034.72 |1043.43 | 1033.09|
| Enable overlap | 1154.43 | 1188.64 | 1175.38 | 1154.65 | 1178.81|**1170.38**|

> Tests were using these cofigs: tp4dp1 enforce-eager num_prompts=48, 1.5k->2k

On A3 platform, enable asnyc exponential overlap will give `Qwen2.5-32B` about 7% performance improvement and will give `Qwen3-32B` about 13.3% performance improvement.

For accuracy-wise, I tested on Math-500 dataset and compare the rating:
| | A3 qwen3-32B tp4dp1 enforce-eager  | A3 qwen2.5-32B tp4dp1 enforce-eager |
| --- | --- | --- |
| Disable overlap | 84.90 | 45.72 |
| Enable overlap | 85.36 | 45.68 |
> Test were using  tempreature=1 seed=1234 output_tokens=10k

Our result shows that enable the async exponential overlap will not introduce huge accuracy drop. 

However, on Qwen3-30B, no matter whether we `enable_expert_parallel` or not, enable asnyc exponential overlap will cause the performance drop:

**Qwen3-30B enable_expert_parallel=True**

| | test1 (tok/s)| test2 (tok/s) | test3 (tok/s) | test4 (tok/s) | test5 (tok/s) | AVG (tok/s) |
| --- | --- | --- | --- | --- | --- | --- |
| Disable overlap | 855.55 | 900.15 | 856.83 | 875.90 | 851.41|867.97|
| Enable overlap | 782.77 | 758.10 | 814.84 | 837.34 |823.42|803.29|

> Tests were using these cofigs: tp4dp1 enforce-eager num_prompts=48, 1.5k->2k

**Qwen3-30B enable_expert_parallel=False**

| | test1 (tok/s)| test2 (tok/s) | test3 (tok/s) | test4 (tok/s) | test5 (tok/s) | AVG (tok/s) |
| --- | --- | --- | --- | --- | --- | --- |
| Disable overlap | 972 | 1025.78 | 1131.05 | 1106.95| 1122.46|1071.78|
| Enable overlap | 1035.24 | 1037.53 | 1087.57 | 1087.56 | 1069.11|1063.40|

> Tests were using these cofigs: tp4dp1 enforce-eager num_prompts=48, 1.5k->2k

Considering the proformance drop in `Qwen3-30B`, we set this feature default to **OFF**, and to enable this feature, we can set `export VLLM_ASCEND_ENABLE_ASYNC_EXPONENTIAL=1`.

- vLLM version: v0.12.0
- vLLM main: https://github.com/vllm-project/vllm/commit/ad32e3e19ccf0526cb6744a5fed09a138a5fb2f9

### CODE REVIEW

File: vllm_ascend/worker/model_runner_v1.py [hunk 1]
```diff
@@ -4480,3 +4490,20 @@ def _to_list(self, sampled_token_ids: torch.Tensor) -> list[np.ndarray]:
...
+        # and overlapping with model executing.
+        with torch.npu.stream(self._async_exponential_stream):
+            self._async_exponential_stream.wait_stream(default_stream)
+            b_s = logits_indices.shape[0]
+            head_dim = self.model_config.get_vocab_size()
+            q = torch.empty((b_s, head_dim), device="npu", dtype=torch.float32)
+            generators = self.input_batch.sampling_metadata.generators
+            if len(generators) != q.shape[0]:
+                q.exponential_()
+            if generators:
```
Comment (jgong5) [inline vllm_ascend/worker/model_runner_v1.py]: what if both `len(generators) != q.shape[0]` and `generators` are `True`?

Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: If both `len(generators) != q.shape[0]` and `generators` are `True`, we just do `q.exponential_()` first, then overwrite each `q[i]` with `q[i].exponential_(generator=generator)`.
This part we simply re-use the same logic in vllm's `random_sample`. Hope this information is helpful!

**Comment by Ronald1995** [comment]
you should add e2e and ut test.

**Comment by YuhanBai** [comment]
> you should add e2e and ut test.

Added a simple e2e test on Qwen3-0.6B, thanks for advice!

File: vllm_ascend/worker/model_runner_v1.py [hunk 2]
```diff
@@ -639,6 +639,16 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
 
         self.transfer_event = torch.npu.Event()
 
+        self.enable_asnyc_exp = envs_ascend.VLLM_ASCEND_ENABLE_ASYNC_EXPONENTIAL
+        if self.enable_asnyc_exp not in [0, 1, 2]:
```
Comment (Ronald1995) [inline vllm_ascend/worker/model_runner_v1.py]: it's better to use `if self.enable_asnyc_exp not in (0, 1, 2):

File: vllm_ascend/worker/model_runner_v1.py [hunk 3]
```diff
@@ -4579,3 +4594,30 @@ def _generate_pcp_mtp_input(
...
+    def _do_async_exponential(self, default_stream, logits_indices):
+        # Calculating exponential randoms in a different stream
+        # and overlapping with model executing.
+        with torch.npu.stream(self._async_exponential_stream):
+            self._async_exponential_stream.wait_stream(default_stream)
+            b_s = logits_indices.shape[0]
+            head_dim = self.model_config.get_vocab_size()
+            q = torch.empty((b_s, head_dim), device="npu", dtype=torch.float32)
+            generators = self.input_batch.sampling_metadata.generators
+            if self.enable_asnyc_exp == 2 and hasattr(torch_npu, "npu_sim_exponential_"):
```
Comment (Ronald1995) [inline vllm_ascend/worker/model_runner_v1.py]: add some comment to explain the magic number 2.

File: vllm_ascend/worker/model_runner_v1.py (hunk 2 above)
Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Fixed

File: vllm_ascend/worker/model_runner_v1.py (hunk 3 above)
Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Added some comments to explain

File: vllm_ascend/worker/model_runner_v1.py [hunk 4]
```diff
@@ -4690,3 +4715,32 @@ def _generate_pcp_mtp_input(
...
+        with torch.npu.stream(self._async_exponential_stream):
+            self._async_exponential_stream.wait_stream(default_stream)
+            b_s = logits_indices.shape[0]
+            head_dim = self.model_config.get_vocab_size()
+            q = torch.empty((b_s, head_dim), device="npu", dtype=torch.float32)
+            generators = self.input_batch.sampling_metadata.generators
+            if self.enable_async_exp == 2:
+                # Set self.enable_async_exp to 2 will enable async exponential with AI-core exponential.
+                # If AI-core exponential is not supported, use the default expontial.
+                if len(generators) != q.shape[0]:
```
Comment (weijinqian0) [inline vllm_ascend/worker/model_runner_v1.py]: to much comment.

File: vllm_ascend/worker/model_runner_v1.py [hunk 5]
```diff
@@ -4690,3 +4715,32 @@ def _generate_pcp_mtp_input(
             self.input_ids_pcp_full_cpu[:total_num_scheduled_tokens_pcp_full],
             non_blocking=True,
         )
+
+    def _do_async_exponential(self, default_stream, logits_indices):
```
Comment (weijinqian0) [inline vllm_ascend/worker/model_runner_v1.py]: why dont move this to sampler?

Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Will move this to sampler in next commit.

File: vllm_ascend/worker/model_runner_v1.py (hunk 4 above)
Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Will fix this in next commit.

File: vllm_ascend/worker/model_runner_v1.py (hunk 5 above)
Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Done

File: vllm_ascend/worker/model_runner_v1.py (hunk 4 above)
Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Done

**Comment by YuhanBai** [comment]
**UPDATE**: 
- Now our overlap will use the same stream that introduced in this pr: #4908 .

- We move the `do_async_exponential` from `model_runner_v1.py` to `sampler.py`.
- Now we are using `additional_config` to enable async exponential:
   - Add `"enable_async_exponential": 1` in `addition_config`.
- Now we **ONLY** support default exponential/AI-CPU exponential, the old `"enable_async_exponential": 2` option has been aborted to keep consistency.
- Solve the conflict introduced by `model_runner_v1` refection.

File: vllm_ascend/worker/model_runner_v1.py [hunk 6]
```diff
@@ -391,6 +391,7 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
         self.reorder_batch_threshold: int | None = None
         self.query_start_loc = self._make_buffer(self.max_num_reqs + 1,
                                                  dtype=torch.int32)
+        self.enable_async_exp = self.ascend_config.enable_async_exponential
```
Comment (zhenwenqi2024) [inline vllm_ascend/worker/model_runner_v1.py]: does vllm has this attr? why we add it here?

Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: We are adding this to control whether we want to enable the async overlap.

File: vllm_ascend/worker/model_runner_v1.py [hunk 7]
```diff
@@ -1508,6 +1509,12 @@ def execute_model(
         aclgraph_runtime_mode, batch_descriptor = \
             self.cudagraph_dispatcher.dispatch(num_tokens=num_input_tokens, uniform_decode=uniform_decode, has_lora=has_lora)
 
+        if self.enable_async_exp != 0:
```
Comment (MengqingCao) [inline vllm_ascend/worker/model_runner_v1.py]: let's directly use `self.ascend_config.enable_async_exponential` instead of adding a new var to modelrunner.
```suggestion
        if self.ascend_config.enable_async_exponential != 0:
```
also cc @zhenwenqi2024

File: vllm_ascend/ascend_config.py [hunk 8]
```diff
@@ -158,6 +158,11 @@ def __init__(self, vllm_config):
                                               False):
             kv_cfg.engine_id = f"{kv_cfg.engine_id}-{uuid4().hex}"
             kv_cfg._engine_id_patched = True
+        self.enable_async_exponential = additional_config.get(
```
Comment (MengqingCao) [inline vllm_ascend/ascend_config.py:164]: plz update the doc on ascend config when adding a new configuration

File: tests/e2e/singlecard/test_sampler.py [hunk 9]
```diff
@@ -47,3 +47,21 @@ def test_models_prompt_logprobs() -> None:
...
+                                     top_k=50,
+                                     top_p=0.9)
+
+    with VllmRunner("Qwen/Qwen3-0.6B",
+                    max_model_len=8192,
+                    gpu_memory_utilization=0.7,
+                    additional_config={
+                        "enable_async_exponential": 1,
+                    }) as runner:
+        runner.generate(example_prompts, sampling_params)
```
Comment (MengqingCao) [inline tests/e2e/singlecard/test_sampler.py:67]: Could we add a golden results here to make sure the accuracy won't be brake

File: vllm_ascend/sample/sampler.py [hunk 10]
```diff
@@ -41,10 +41,33 @@ def __init__(self, logprobs_mode=DEFAULT_LOGPROBS_MODE):
...
+            if generators:
+                for i, generator in generators.items():
+                    q[i].exponential_(generator=generator)
+            self.async_exponential_event.record()
+        self.set_q_event(q, self.async_exponential_event)
 
 
 class AscendTopKTopPSampler(TopKTopPSampler):
 
+    def set_q_event(self, q, event):
```
Comment (MengqingCao) [inline vllm_ascend/sample/sampler.py:68]: plz add a comment here to describe this function

File: tests/e2e/singlecard/test_sampler.py (hunk 9 above)
Comment (YuhanBai) [reply tests/e2e/singlecard/test_sampler.py:67]: In this pr I only added a simple e2e test to ensure my change won't break the whole process.
For accuracy-wise, I tested my change on Qwen3-30B/Qwen2.5-32B/Qwen3-32B on Math-500 dataset. And result showed that my change won't affect the accuracy.
If necessary, I will add some more tests in future.

File: vllm_ascend/sample/sampler.py (hunk 10 above)
Comment (YuhanBai) [reply vllm_ascend/sample/sampler.py:68]: Done

File: vllm_ascend/ascend_config.py (hunk 8 above)
Comment (YuhanBai) [reply vllm_ascend/ascend_config.py:164]: Done

File: vllm_ascend/worker/model_runner_v1.py (hunk 7 above)
Comment (YuhanBai) [reply vllm_ascend/worker/model_runner_v1.py]: Done

File: vllm_ascend/ascend_config.py [hunk 11]
```diff
@@ -166,6 +166,11 @@ def __init__(self, vllm_config):
...
+            "enable_async_exponential", 0)
+        if self.enable_async_exponential not in (0, 1):
+            raise AssertionError(
```
Comment (zhenwenqi2024) [inline vllm_ascend/ascend_config.py:167]: Can we use Bool for the judgment here?

File: vllm_ascend/sample/sampler.py [hunk 12]
```diff
@@ -99,4 +124,8 @@ def forward_native(self, logits, generators, k, p):
             logits_to_return = logits.log_softmax(dim=-1, dtype=torch.float32)
 
         probs = logits.softmax(dim=-1, dtype=torch.float32)
+        if getattr(self, "q", None) is not None:
```
Comment (zhenwenqi2024) [inline vllm_ascend/sample/sampler.py]: getattr has pool performance, we should avoid this

File: vllm_ascend/ascend_config.py (hunk 11 above)
Comment (YuhanBai) [reply vllm_ascend/ascend_config.py:167]: We use `int` here is for a potential AI-Core exponential support, our initial thought is that set to `2` will enable AI-Core exponential overlap, which has different API. We removed this part in this PR to maintain the consistency. In future we will add that support, once this PR is merged.

File: vllm_ascend/sample/sampler.py (hunk 12 above)
Comment (YuhanBai) [reply vllm_ascend/sample/sampler.py]: I think it should be ok, even though using `getattr` might have performance issue, our performance gain will cover that part.

File: tests/e2e/singlecard/test_sampler.py (hunk 9 above)
Comment (MengqingCao) [reply tests/e2e/singlecard/test_sampler.py:67]: When not using greedy sample, maybe we couldn't get a golden result here. You can have a try locally and I'll keep it as is currently

File: vllm_ascend/sample/sampler.py (hunk 12 above)
Comment (MengqingCao) [reply vllm_ascend/sample/sampler.py]: maybe we can just set the default `self.q=None` to avoid this

Comment (YuhanBai) [reply vllm_ascend/sample/sampler.py]: Update: now we are using enable_async_exponential config to decide goes to which line of code.

**Comment by Yikun** [comment]
Thanks for your first contributions! Your awesome first PR has been included in vLLM Ascend v0.13.0rc1 release.

And 👍 for your commit message and test cover.

[1] https://github.com/vllm-project/vllm-ascend/releases/tag/v0.13.0rc1
[2] https://mp.weixin.qq.com/s/3Psz3mYFTLktgSEDGqM9wQ
